pyteal/ast/router.py

```diff
@@ -0,0 +1,471 @@
+        Returns:
+            Expr:
+                - for bare-appcall it returns an expression that the handler takes no txn arg and Approve
+                - for abi-method it returns the txn args correctly decomposed into ABI variables,
+                  passed in ABIReturnSubroutine and logged, then approve.
+        """
+        if not is_method_call:
+            match handler:
+                case Expr():
+                    if handler.type_of() != TealType.none:
+                        raise TealInputError(
+                            "bare appcall handler should have type to be none."
+                        )
+                    return handler if handler.has_return() else Seq(handler, Approve())
+                case SubroutineFnWrapper():
+                    if handler.type_of() != TealType.none:
+                        raise TealInputError(
+                            f"subroutine call should be returning none not {handler.type_of()}."
+                        )
+                    if handler.subroutine.argument_count() != 0:
+                        raise TealInputError(
+                            f"subroutine call should take 0 arg for bare-app call. "
+                            f"this subroutine takes {handler.subroutine.argument_count()}."
+                        )
+                    return Seq(handler(), Approve())
+                case ABIReturnSubroutine():
+                    if handler.type_of() != "void":
+                        raise TealInputError(
+                            f"abi-returning subroutine call should be returning void not {handler.type_of()}."
+                        )
+                    if handler.subroutine.argument_count() != 0:
+                        raise TealInputError(
+                            f"abi-returning subroutine call should take 0 arg for bare-app call. "
+                            f"this abi-returning subroutine takes {handler.subroutine.argument_count()}."
+                        )
+                    return Seq(cast(Expr, handler()), Approve())
+                case _:
+                    raise TealInputError(
+                        "bare appcall can only accept: none type Expr, or Subroutine/ABIReturnSubroutine with none return and no arg"
+                    )
+        else:
+            if not isinstance(handler, ABIReturnSubroutine):
+                raise TealInputError(
+                    f"method call should be only registering ABIReturnSubroutine, got {type(handler)}."
+                )
+            if not handler.is_registrable():
+                raise TealInputError(
+                    f"method call ABIReturnSubroutine is not registrable"
+                    f"got {handler.subroutine.argument_count()} args with {len(handler.subroutine.abi_args)} ABI args."
+                )
+
+            arg_type_specs: list[abi.TypeSpec] = cast(
```

Comment:
Does the type hint here do anything the cast doesn't?

Reply:
no, it is just added here for my own eye reading... removable and removing

Reply:
Mostly curious, not a knock